akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -414,13 +459,17 @@ private[stream] object ConnectionSourceStage {
       new InHandler {
         override def onPush(): Unit = {
           val elem = grab(bytesIn)
+          elemCount += 1
           ReactiveStreamsCompliance.requireNonNullElement(elem)
           if (writeInProgress) {
             writeBuffer = writeBuffer ++ elem
+          } else if (!smartBatching || writeBuffer.size >= writeBufferSize) {
+            writeBuffer = writeBuffer ++ elem
+            sendWriteBuffer()
           } else {
-            connection ! Write(writeBuffer ++ elem, WriteAck)
-            writeInProgress = true
-            writeBuffer = ByteString.empty
+            writeBuffer = writeBuffer ++ elem
+            writeDelayCountDown = 20
+            sendWriteDelay()
```

Comment:
@jrudolph I have changed this so that it tries to batch also for the "first" element. That improves it further, and we don't introduce much extra latency in case it's a single element message.

Logs to compare with and without the smart batching enabled when running locally:

```
ghz --insecure --call perf.action.PingActionService/Ping -d '{}' -n 1000000 --concurrency=1000 --connections=2 --skipFirst=1000 localhost:9000

SMART_BATCHING=false

proxy:
[info] # [1725088177] writeBuffer size 17, avg 213 bytes/write, elemCount 1229190, writeCount 75000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 117, avg 213 bytes/write, elemCount 1243305, writeCount 76000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 15, avg 213 bytes/write, elemCount 1245431, writeCount 76000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 219, avg 213 bytes/write, elemCount 1259897, writeCount 77000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 17, avg 213 bytes/write, elemCount 1258721, writeCount 77000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 414, avg 213 bytes/write, elemCount 1274228, writeCount 78000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 24, avg 213 bytes/write, elemCount 1274549, writeCount 78000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 78, avg 212 bytes/write, elemCount 1288989, writeCount 79000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 117, avg 213 bytes/write, elemCount 1289925, writeCount 79000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 17, avg 212 bytes/write, elemCount 1305257, writeCount 80000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 17, avg 212 bytes/write, elemCount 1305860, writeCount 80000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 17, avg 212 bytes/write, elemCount 1321102, writeCount 81000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 39, avg 213 bytes/write, elemCount 1324381, writeCount 81000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 17, avg 212 bytes/write, elemCount 1335088, writeCount 82000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 17, avg 212 bytes/write, elemCount 1336961, writeCount 82000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 17, avg 212 bytes/write, elemCount 1349223, writeCount 83000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 446, avg 212 bytes/write, elemCount 1353378, writeCount 83000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 15, avg 211 bytes/write, elemCount 1363195, writeCount 84000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 141, avg 212 bytes/write, elemCount 1366825, writeCount 84000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 563, avg 211 bytes/write, elemCount 1376263, writeCount 85000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 429, avg 212 bytes/write, elemCount 1381379, writeCount 85000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 39, avg 211 bytes/write, elemCount 1392554, writeCount 86000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 17, avg 212 bytes/write, elemCount 1398239, writeCount 86000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 15, avg 211 bytes/write, elemCount 1407362, writeCount 87000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 39, avg 211 bytes/write, elemCount 1412354, writeCount 87000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 273, avg 210 bytes/write, elemCount 1419711, writeCount 88000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 156, avg 211 bytes/write, elemCount 1425351, writeCount 88000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 212, avg 210 bytes/write, elemCount 1435103, writeCount 89000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 15, avg 211 bytes/write, elemCount 1439835, writeCount 89000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 15, avg 210 bytes/write, elemCount 1453520, writeCount 90000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 156, avg 211 bytes/write, elemCount 1459025, writeCount 90000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 453, avg 210 bytes/write, elemCount 1471285, writeCount 91000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 819, avg 211 bytes/write, elemCount 1475045, writeCount 91000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 17, avg 210 bytes/write, elemCount 1487724, writeCount 92000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 63, avg 211 bytes/write, elemCount 1491762, writeCount 92000, delayCount 0, writeDelayCountDown 0
[info] # [1725088177] writeBuffer size 290, avg 211 bytes/write, elemCount 1506701, writeCount 93000, delayCount 0, writeDelayCountDown 0
[info] # [309790461] writeBuffer size 312, avg 211 bytes/write, elemCount 1509738, writeCount 93000, delayCount 0, writeDelayCountDown 0

user-function:
# [-1198256216] writeBuffer size 5696, avg 5649 bytes/write, elemCount 3046953, writeCount 16000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 15, avg 5630 bytes/write, elemCount 3226571, writeCount 17000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5696, avg 5625 bytes/write, elemCount 3413453, writeCount 18000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 10947, avg 5644 bytes/write, elemCount 3615060, writeCount 19000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 2848, avg 5666 bytes/write, elemCount 3819950, writeCount 20000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 7565, avg 5669 bytes/write, elemCount 4013361, writeCount 21000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5429, avg 5681 bytes/write, elemCount 4212982, writeCount 22000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 1513, avg 5704 bytes/write, elemCount 4422866, writeCount 23000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 534, avg 5694 bytes/write, elemCount 4607073, writeCount 24000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5592, avg 5676 bytes/write, elemCount 4783769, writeCount 25000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5785, avg 5653 bytes/write, elemCount 4954998, writeCount 26000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5770, avg 5640 bytes/write, elemCount 5133526, writeCount 27000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 15, avg 5625 bytes/write, elemCount 5309862, writeCount 28000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 15, avg 5608 bytes/write, elemCount 5482162, writeCount 29000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5592, avg 5613 bytes/write, elemCount 5676409, writeCount 30000, delayCount 0, writeDelayCountDown 0
# [-1198256216] writeBuffer size 5696, avg 5626 bytes/write, elemCount 5879747, writeCount 31000, delayCount 0, writeDelayCountDown 0

SMART_BATCHING=true

proxy:
[info] # [-1296688204] writeBuffer size 134, avg 669 bytes/write, elemCount 1027263, writeCount 20000, delayCount 57203, writeDelayCountDown 18
[info] # [1661118089] writeBuffer size 524, avg 671 bytes/write, elemCount 1030265, writeCount 20000, delayCount 57643, writeDelayCountDown 15
[info] # [-1296688204] writeBuffer size 39, avg 641 bytes/write, elemCount 1033953, writeCount 21000, delayCount 59307, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 39, avg 643 bytes/write, elemCount 1036790, writeCount 21000, delayCount 59700, writeDelayCountDown 18
[info] # [-1296688204] writeBuffer size 273, avg 637 bytes/write, elemCount 1076887, writeCount 22000, delayCount 62189, writeDelayCountDown 18
[info] # [1661118089] writeBuffer size 39, avg 640 bytes/write, elemCount 1081949, writeCount 22000, delayCount 62540, writeDelayCountDown 18
[info] # [-1296688204] writeBuffer size 234, avg 642 bytes/write, elemCount 1134373, writeCount 23000, delayCount 65112, writeDelayCountDown 16
[info] # [1661118089] writeBuffer size 39, avg 645 bytes/write, elemCount 1138801, writeCount 23000, delayCount 65601, writeDelayCountDown 18
[info] # [1661118089] writeBuffer size 195, avg 648 bytes/write, elemCount 1193860, writeCount 24000, delayCount 68419, writeDelayCountDown 19
[info] # [-1296688204] writeBuffer size 39, avg 648 bytes/write, elemCount 1193869, writeCount 24000, delayCount 67981, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 17, avg 646 bytes/write, elemCount 1241494, writeCount 25000, delayCount 71296, writeDelayCountDown 19
[info] # [-1296688204] writeBuffer size 2240, avg 647 bytes/write, elemCount 1241652, writeCount 25000, delayCount 70939, writeDelayCountDown 14
[info] # [1661118089] writeBuffer size 78, avg 647 bytes/write, elemCount 1292381, writeCount 26000, delayCount 74334, writeDelayCountDown 19
[info] # [-1296688204] writeBuffer size 702, avg 646 bytes/write, elemCount 1290381, writeCount 26000, delayCount 73849, writeDelayCountDown 13
[info] # [1661118089] writeBuffer size 78, avg 652 bytes/write, elemCount 1351505, writeCount 27000, delayCount 77235, writeDelayCountDown 18
[info] # [-1296688204] writeBuffer size 17, avg 652 bytes/write, elemCount 1352082, writeCount 27000, delayCount 76728, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 78, avg 653 bytes/write, elemCount 1404199, writeCount 28000, delayCount 80095, writeDelayCountDown 19
[info] # [-1296688204] writeBuffer size 234, avg 653 bytes/write, elemCount 1404069, writeCount 28000, delayCount 79632, writeDelayCountDown 14
[info] # [1661118089] writeBuffer size 2591, avg 647 bytes/write, elemCount 1442227, writeCount 29000, delayCount 82957, writeDelayCountDown 17
[info] # [-1296688204] writeBuffer size 17, avg 646 bytes/write, elemCount 1440141, writeCount 29000, delayCount 82480, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 212, avg 630 bytes/write, elemCount 1452565, writeCount 30000, delayCount 85163, writeDelayCountDown 18
[info] # [-1296688204] writeBuffer size 17, avg 629 bytes/write, elemCount 1448963, writeCount 30000, delayCount 84708, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 134, avg 612 bytes/write, elemCount 1457931, writeCount 31000, delayCount 87084, writeDelayCountDown 16
[info] # [-1296688204] writeBuffer size 39, avg 611 bytes/write, elemCount 1455812, writeCount 31000, delayCount 86725, writeDelayCountDown 19
[info] # [1661118089] writeBuffer size 914, avg 615 bytes/write, elemCount 1511913, writeCount 32000, delayCount 89829, writeDelayCountDown 17

user-function:
# [19246132] writeBuffer size 7728, avg 12938 bytes/write, elemCount 3052788, writeCount 7000, delayCount 17531, writeDelayCountDown 16
# [19246132] writeBuffer size 8544, avg 12957 bytes/write, elemCount 3494049, writeCount 8000, delayCount 20007, writeDelayCountDown 17
# [19246132] writeBuffer size 16391, avg 12996 bytes/write, elemCount 3942660, writeCount 9000, delayCount 22427, writeDelayCountDown 17
# [19246132] writeBuffer size 16440, avg 12987 bytes/write, elemCount 4377802, writeCount 10000, delayCount 24965, writeDelayCountDown 16
# [19246132] writeBuffer size 16391, avg 13016 bytes/write, elemCount 4826223, writeCount 11000, delayCount 27430, writeDelayCountDown 17
# [19246132] writeBuffer size 14848, avg 13026 bytes/write, elemCount 5269210, writeCount 12000, delayCount 29932, writeDelayCountDown 17
# [19246132] writeBuffer size 7387, avg 13060 bytes/write, elemCount 5723335, writeCount 13000, delayCount 32399, writeDelayCountDown 17
```

From this you can see that the writeDelayCountDown sometimes uses more than 1 roundtrip.

Reply:
Great, thanks for testing.

Reply:
We should probably make it configurable? The result here is just the result of a particular "fairness conditions", i.e. the amount of parallel work in comparison to the number of threads available.

Could there be other conditions where those 20 round trips could lead to more excessive delays?

If your system is CPU-constrained, latency is introduced by work being queued waiting for CPU, so in that case batching should always help when it can reduce average CPU usage for a task (here by amortizing system call costs over more bytes to send). So, probably no danger from that perspective. Not sure if there could be other scenarios where it could do more harm.

Reply:
I tried this in the staging environment and it works as expected. Not batching as much as in local. user function avg around 300 bytes/write, and proxy around 150 bytes/write.

> We should probably make it configurable?

Yes, we should. Do you think it should be one single config property for the ActorSystem (reference.conf) or should it be some kind of Streams attribute? If we make it an attribute it should anyway have a config property in Akka HTTP and Artery. I'm leaning towards keeping it simple and just have a single config property.

I think we can have a default of 10.

Reply:
I agree, there's no need to overcomplicate and we could still later introduce an attribute that takes a default from a setting if needed.